server/src/main/java/org/apache/iotdb/db/metadata/MTree.java

```diff
@@ -762,28 +762,47 @@ private void findStorageGroup(MNode node, String path, List<String> res) {
    *
    * @return a list contains all distinct devices
    */
-  Set<String> getAllDevices() {
-    HashSet<String> devices = new HashSet<>();
-    MNode node;
-    if ((node = getRoot()) != null) {
-      findDevices(node, SQLConstant.ROOT, devices);
+  Set<String> getAllDevices() throws PathException {
+    return getDevices(SQLConstant.ROOT);
+  }
+
+  /**
+   * Get all devices in current Metadata Tree with prefixPath.
+   *
+   * @return a list contains all distinct devices
+   */
+  Set<String> getDevices(String prefixPath) throws PathException {
+    String[] nodes = MetaUtils.getNodeNames(prefixPath, PATH_SEPARATOR);
+    if (nodes.length == 0 || !nodes[0].equals(getRoot().getName())) {
+      throw new MTreePathException("PrefixPath", prefixPath);
     }
+    HashSet<String> devices = new HashSet<>();
+    findDevices(getRoot(), nodes, 1, "", devices);
     return new LinkedHashSet<>(devices);
   }
 
-  private void findDevices(MNode node, String path, HashSet<String> res) {
-    if (node == null) {
-      return;
-    }
-    if (node.isLeaf()) {
-      res.add(path);
-      return;
+  private void findDevices(MNode node, String[] nodes, int idx, String parent, HashSet<String> res) {
+    String nodeReg;
+    if (idx >= nodes.length) {
+      nodeReg = "*";
+    } else {
+      nodeReg = nodes[idx];
     }
-    for (MNode child : node.getChildren().values()) {
-      if (child.isLeaf()) {
-        res.add(path);
-      } else {
-        findDevices(child, path + "." + child.toString(), res);
+    if (!("*").equals(nodeReg)) {
+      if (node.hasChild(nodeReg)) {
+        if(node.getChild(nodeReg).isLeaf()){
+          res.add(parent + node.getName());
+        } else{
+          findDevices(node.getChild(nodeReg), nodes, idx + 1, parent + node.getName() + ".", res);
+        }
+      }
+    } else {
+      for (MNode child : node.getChildren().values()) {
+        if(child.isLeaf()){
+          res.add(parent + node.getName());
```

Comment:
You may set a flag here so you won't need to add the same device multiple times.

Reply:
Oh! I forgot that. Thank you. I will fix it.